// neutrinos-oauth-client/esm5/lib/neutrinos-oauth-client.module.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/neutrinos-oauth-client.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule, APP_INITIALIZER } from '@angular/core';
import { NeutrinosOauthComponent } from './neutrinos-oauth.component';
import { HttpClientModule, HTTP_INTERCEPTORS } from '@angular/common/http';
import { CookieInterceptorService } from './cookie-interceptor.service';
import { NeutrinosOAuthClientService } from './neutrinos-oauth-client.service';
import { UtilService } from './services/util.service';
/**
 * @param {?} neutrinosOAuthService
 * @param {?} utilService
 * @return {?}
 */
export function getTokens(neutrinosOAuthService, utilService) {
    return (/**
     * @return {?}
     */
    function () {
        /**
         * For ios restore the cookies and get userInfo since iOS webview does not
         * persisit the cookie between app launches
         */
        if (utilService.getEnvValue('isIDSEnabled') === 'true') {
            if (window['device'] && window['device'].platform === 'iOS') {
                return neutrinosOAuthService.restoreCookies().then((/**
                 * @param {?} cookieExist
                 * @return {?}
                 */
                function (cookieExist) {
                    console.log('cookieExist' + cookieExist);
                    cookieExist ? neutrinosOAuthService.getUserInfo()
                        .catch((/**
                     * @param {?} userInfoErr
                     * @return {?}
                     */
                    function (userInfoErr) {
                        console.log(userInfoErr);
                    })) : null;
                }));
            }
            else {
                // Cannot get cookie for cross domain using JS. hence calling userinfo always
                return neutrinosOAuthService.getUserInfo()
                    .catch((/**
                 * @param {?} userInfoErr
                 * @return {?}
                 */
                function (userInfoErr) {
                    console.log(userInfoErr);
                }));
            }
        }
    });
}
// @dynamic (https://github.com/ng-packagr/ng-packagr/issues/641)
var NeutrinosOAuthClientModule = /** @class */ (function () {
    function NeutrinosOAuthClientModule() {
    }
    NeutrinosOAuthClientModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [NeutrinosOauthComponent],
                    imports: [
                        HttpClientModule
                    ],
                    providers: [
                        {
                            provide: HTTP_INTERCEPTORS,
                            useClass: CookieInterceptorService,
                            multi: true
                        },
                        {
                            provide: APP_INITIALIZER,
                            useFactory: getTokens,
                            deps: [NeutrinosOAuthClientService, UtilService],
                            multi: true
                        },
                    ]
                },] }
    ];
    return NeutrinosOAuthClientModule;
}());
export { NeutrinosOAuthClientModule };
