// neutrinos-oauth-client/esm5/lib/neutrinos-oauth.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/neutrinos-oauth.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component } from '@angular/core';
import { NeutrinosOAuthClientService } from './neutrinos-oauth-client.service';
import { HttpClient } from '@angular/common/http';
var NeutrinosOauthComponent = /** @class */ (function () {
    function NeutrinosOauthComponent(neutrinosOauthService, http) {
        this.neutrinosOauthService = neutrinosOauthService;
        this.http = http;
    }
    /**
     * @return {?}
     */
    NeutrinosOauthComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.neutrinosOauthService.authState().subscribe((/**
         * @param {?} state
         * @return {?}
         */
        function (state) {
            console.log(state);
        }));
    };
    /**
     * @return {?}
     */
    NeutrinosOauthComponent.prototype.login = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var a = this.neutrinosOauthService.login('/i-am-loggedin').then((/**
         * @param {?} tokens
         * @return {?}
         */
        function (tokens) {
            console.log('got tokens in component', tokens);
        })).catch((/**
         * @param {?} e
         * @return {?}
         */
        function (e) {
            console.error(e);
        }));
        // console.log(a);
    };
    /**
     * @return {?}
     */
    NeutrinosOauthComponent.prototype.token = /**
     * @return {?}
     */
    function () {
        this.neutrinosOauthService.getUserInfo().then((/**
         * @param {?} t
         * @return {?}
         */
        function (t) {
            console.log('t', t);
        })).catch(console.log);
    };
    /**
     * @return {?}
     */
    NeutrinosOauthComponent.prototype.logout = /**
     * @return {?}
     */
    function () {
        this.neutrinosOauthService.logout();
    };
    /**
     * @return {?}
     */
    NeutrinosOauthComponent.prototype.userinfo = /**
     * @return {?}
     */
    function () {
        alert(JSON.stringify(this.neutrinosOauthService.userInfo));
        console.log(this.neutrinosOauthService.userInfo);
    };
    /**
     * @return {?}
     */
    NeutrinosOauthComponent.prototype.httpTest = /**
     * @return {?}
     */
    function () {
        this.http.get('http://localhost:8081/api/ping').subscribe((/**
         * @param {?} res
         * @return {?}
         */
        function (res) {
            console.log(res);
        }), (/**
         * @param {?} e
         * @return {?}
         */
        function (e) {
            console.error(e);
        }));
    };
    NeutrinosOauthComponent.decorators = [
        { type: Component, args: [{
                    selector: 'lib-neutrinos-oauth',
                    template: "\n   NOTE: use this component only for testing. Not intended for production\n\n      <button (click)=\"login()\" >  Login </button>\n      <button (click)=\"token()\" > Get tokenset</button>\n      <button (click)=\"logout()\" > Logout </button>\n      <button (click)=\"userinfo()\" > In memory user info </button>\n      <button (click)=\"httpTest()\" > Http Interceptor test </button>\n      "
                }] }
    ];
    /** @nocollapse */
    NeutrinosOauthComponent.ctorParameters = function () { return [
        { type: NeutrinosOAuthClientService },
        { type: HttpClient }
    ]; };
    return NeutrinosOauthComponent;
}());
export { NeutrinosOauthComponent };
if (false) {
    /** @type {?} */
    NeutrinosOauthComponent.prototype.neutrinosOauthService;
    /**
     * @type {?}
     * @private
     */
    NeutrinosOauthComponent.prototype.http;
}
