// neutrinos-module/esm5/src/app/neutrinos-module/neutrinos-components/nToggleoptionsComponent/n-toggleoptions.component.js
import { __decorate } from "tslib";
// import { MatSnackBar } from '@angular/material';
import { Component, Input, EventEmitter, Output } from '@angular/core';
var NToggleoptionsComponent = /** @class */ (function () {
    function NToggleoptionsComponent() {
        // options = {icon: 'home', value: 'home', disabled: false, 'checked': false}
        this.toggleOptions = [];
        this.align = 'horizontal';
        this.indexChange = new EventEmitter();
        this.valueChange = new EventEmitter();
    }
    // constructor(private snackbar: MatSnackBar) {
    // }
    NToggleoptionsComponent.prototype.ngOnInit = function () {
        if (this.checkNumber(this.checkedIndex)) {
            this.checkedIndex = Number(this.checkedIndex);
        }
        if (this.checkNumber(this.disabledIndex)) {
            this.disabledIndex = Number(this.disabledIndex);
        }
    };
    NToggleoptionsComponent.prototype.ngOnChanges = function (changes) {
        if (changes['checkedIndex'] && this.checkNumber(this.checkedIndex)) {
            this.checkedIndex = Number(this.checkedIndex);
        }
        else if (changes['disabledIndex'] && this.checkNumber(this.disabledIndex)) {
            this.disabledIndex = Number(this.disabledIndex);
        }
    };
    NToggleoptionsComponent.prototype.optionClicked = function (index) {
        if (this.toggleOptions && this.toggleOptions.length > 0) {
            this.indexChange.emit(index);
            if (this.toggleOptions[index] && this.toggleOptions[index].value) {
                this.valueChange.emit(this.toggleOptions[index].value);
            }
            else {
                // this.snackbar.open('Invalid toggle button value', 'OK');
            }
        }
        else {
            // this.snackbar.open('Invalid toggle options', 'OK');
        }
    };
    NToggleoptionsComponent.prototype.checkNumber = function (number) {
        if (number !== undefined && number !== null) {
            return Number.isNaN(Number(number));
        }
        return false;
    };
    __decorate([
        Input('toggleOptions')
    ], NToggleoptionsComponent.prototype, "toggleOptions", void 0);
    __decorate([
        Input('align')
    ], NToggleoptionsComponent.prototype, "align", void 0);
    __decorate([
        Input('disabledIndex')
    ], NToggleoptionsComponent.prototype, "disabledIndex", void 0);
    __decorate([
        Input('checkedIndex')
    ], NToggleoptionsComponent.prototype, "checkedIndex", void 0);
    __decorate([
        Input('value')
    ], NToggleoptionsComponent.prototype, "value", void 0);
    __decorate([
        Output()
    ], NToggleoptionsComponent.prototype, "indexChange", void 0);
    __decorate([
        Output()
    ], NToggleoptionsComponent.prototype, "valueChange", void 0);
    NToggleoptionsComponent = __decorate([
        Component({
            selector: 'n-toggleoptions',
            template: "<mat-button-toggle-group #group=\"matButtonToggleGroup\" [vertical]=\"align == 'vertical'? true: false\" [value]=\"value\">\r\n    <mat-button-toggle *ngFor=\"let option of toggleOptions; let i = index;\" [value]=\"option.value\" [disabled]=\"option.disabled || disabledIndex == i\" [checked]=\"option.checked || checkedIndex == i\" (click)=\"optionClicked(i)\" fxLayoutAlign=\"center center\">\r\n        <mat-icon *ngIf=\"option.icon\">{{option.icon}}</mat-icon>\r\n        <span *ngIf=\"!option.icon && option.value\">{{option.value}}</span>\r\n    </mat-button-toggle>\r\n</mat-button-toggle-group>"
        })
    ], NToggleoptionsComponent);
    return NToggleoptionsComponent;
}());
export { NToggleoptionsComponent };
