// neutrinos-seed-services/esm5/src/app/neutrinos-seed-services/n-HTTPLoader.service.js
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
// import { NSnackbarService } from './n-snackbar.service';
import { NLogoutService } from './n-logout.service';
import { NPubSubService } from './n-pubSub.service';
// import { NSnackbarService } from './snackbar/service/snackbar.service';
var NHTTPLoaderService = /** @class */ (function () {
    function NHTTPLoaderService(nLogoutService, nPubSubService) {
        this.nLogoutService = nLogoutService;
        this.nPubSubService = nPubSubService;
        this.pendingRequests = 0;
        this.showLoading = false;
        this._isHTTPRequestInProgress = new Subject();
        this._isHTTPRequestInProgress$ = this._isHTTPRequestInProgress.asObservable();
    }
    NHTTPLoaderService.prototype.isHTTPRequestInProgress = function (bool) {
        if (bool) {
            this.pendingRequests++;
            if (!this.showLoading) {
                this._isHTTPRequestInProgress.next(bool);
            }
            this.showLoading = true;
        }
        else {
            this.pendingRequests--;
            if (this.pendingRequests <= 0) {
                if (this.showLoading) {
                    this._isHTTPRequestInProgress.next(bool);
                }
                this.showLoading = false;
            }
        }
    };
    NHTTPLoaderService.prototype.alertError = function (error) {
        if (error.status < 200 || error.status > 500) {
            if (error.status === 0) {
                this.nPubSubService.$pub('openSnackBar', 'Connectivity issue');
            }
            else {
                this.nPubSubService.$pub('openSnackBar', 'Response failure');
            }
        }
        else if (error.status === 401) {
            this.nLogoutService.logout();
            if (error.error) {
                this.nPubSubService.$pub('openSnackBar', error.error);
            }
            else {
                this.nPubSubService.$pub('openSnackBar', error);
            }
        }
        else if (error.error) {
            if (error.error instanceof Object) {
                this.nPubSubService.$pub('openSnackBar', error.error.message);
            }
            else {
                this.nPubSubService.$pub('openSnackBar', error.error);
            }
        }
        else {
            this.nPubSubService.$pub('openSnackBar', 'Response failure');
        }
    };
    NHTTPLoaderService.ctorParameters = function () { return [
        { type: NLogoutService },
        { type: NPubSubService }
    ]; };
    NHTTPLoaderService = __decorate([
        Injectable()
    ], NHTTPLoaderService);
    return NHTTPLoaderService;
}());
export { NHTTPLoaderService };
