// neutrinos-seed-services/esm5/src/app/neutrinos-seed-services/n-snackbar.service.js
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
// import { MatSnackBar } from '@angular/material';
import { Subject } from 'rxjs';
import { NSystemService } from './n-system.service';
import { NPubSubService } from './n-pubSub.service';
var NSnackbarService = /** @class */ (function () {
    function NSnackbarService(nPubSubService) {
        var _this = this;
        this.nPubSubService = nPubSubService;
        this.snackService = new Subject();
        if (NSystemService.getInstance().getVal('useDefaultExceptionUI')) {
            this.snackBarListner = this.nPubSubService.$sub('openSnackBar', function (msg) {
                _this.openSnackBar(msg);
            });
        }
    }
    NSnackbarService.prototype.get = function () {
        return this.snackService.asObservable();
    };
    NSnackbarService.prototype.add = function (data) {
        this.snackService.next({
            action: 'add',
            data: data
        });
    };
    NSnackbarService.prototype.remove = function (id) {
        this.snackService.next({ action: 'remove', id: id });
    };
    NSnackbarService.prototype.clear = function () {
        this.snackService.next({ action: 'clear' });
    };
    NSnackbarService.prototype.openSnackBar = function (message, timeout) {
        this.add({
            msg: message,
            timeout: timeout || 5000,
            action: {
                text: 'Close',
                onClick: function (snack) {
                },
            },
            onAdd: function (snack) {
            },
            onRemove: function (snack) {
            }
        });
    };
    NSnackbarService.ctorParameters = function () { return [
        { type: NPubSubService }
    ]; };
    NSnackbarService = __decorate([
        Injectable()
    ], NSnackbarService);
    return NSnackbarService;
}());
export { NSnackbarService };
