// neutrinos-seed-services/esm5/src/app/neutrinos-seed-services/neutrinos-seed-services.module.js
import { __decorate } from "tslib";
import { NgModule, APP_INITIALIZER } from '@angular/core';
import { CommonModule } from '@angular/common';
import { NAuthGuardService } from './n-authGaurd.service';
import { NDataModelService } from './n-dataModel.service';
import { NHttpService } from './n-HTTP.service';
import { NHTTPLoaderService } from './n-HTTPLoader.service';
import { NLocalStorageService } from './n-localStorage.service';
import { NLoginService } from './n-login.service';
import { NLogoutService } from './n-logout.service';
import { NNotificationService } from './n-notification.service';
import { NSnackbarService } from './n-snackbar.service';
import { NPubSubService } from './n-pubSub.service';
import { NSessionStorageService } from './n-sessionStorage.service';
import { NTokenService } from './n-token.service';
import { HttpClientModule, HTTP_INTERCEPTORS } from '@angular/common/http';
import { RouterModule } from '@angular/router';
import { InstanceFactory, NgForage, NgForageConfig } from 'ngforage';
import { SnackComponent } from './snackbar/snack/snack.component';
import { SnackbarComponent } from './snackbar/snackbar/snackbar.component';
export function startupServiceFactory(startupService) {
    var res = function () {
        // console.log('module before initstorage call');
        return startupService.initStorage();
    };
    return res;
}
export function NgForageFactory(ngForageConfig) {
    // @ts-ignore
    return new NgForage({}, new InstanceFactory(ngForageConfig));
}
var NeutrinosSeedServicesModule = /** @class */ (function () {
    function NeutrinosSeedServicesModule() {
    }
    NeutrinosSeedServicesModule = __decorate([
        NgModule({
            imports: [
                CommonModule,
                HttpClientModule,
                RouterModule
            ],
            providers: [
                {
                    provide: HTTP_INTERCEPTORS,
                    useClass: NHttpService,
                    multi: true
                },
                {
                    // Provider for APP_INITIALIZER
                    provide: APP_INITIALIZER,
                    useFactory: startupServiceFactory,
                    deps: [NLocalStorageService],
                    multi: true
                },
                {
                    provide: NgForage,
                    useFactory: NgForageFactory,
                    deps: [
                        NgForageConfig
                    ]
                },
                NAuthGuardService,
                NDataModelService,
                NHttpService,
                NHTTPLoaderService,
                NLocalStorageService,
                NLoginService,
                NLogoutService,
                NNotificationService,
                NSnackbarService,
                NPubSubService,
                NSessionStorageService,
                NTokenService
            ],
            declarations: [
                SnackComponent,
                SnackbarComponent
            ],
            exports: [
                SnackComponent,
                SnackbarComponent
            ]
        })
    ], NeutrinosSeedServicesModule);
    return NeutrinosSeedServicesModule;
}());
export { NeutrinosSeedServicesModule };
