// neutrinos-seed-services/esm2015/src/app/neutrinos-seed-services/n-sessionStorage.service.js
var NSessionStorageService_1;
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
let NSessionStorageService = NSessionStorageService_1 = class NSessionStorageService {
    constructor() {
        NSessionStorageService_1.sessionStorageCache = sessionStorage;
    }
    getSessionStorage() {
        return NSessionStorageService_1.sessionStorageCache;
    }
    setValue(key, value) {
        NSessionStorageService_1.sessionStorageCache[key] = value;
        sessionStorage.setItem(key, value);
    }
    getValue(key) {
        if (!NSessionStorageService_1.sessionStorageCache[key]) {
            return null;
        }
        try {
            const obj = NSessionStorageService_1.sessionStorageCache[key];
            return JSON.parse(obj);
        }
        catch (error) {
            return NSessionStorageService_1.sessionStorageCache[key];
        }
    }
    remove(key) {
        if (NSessionStorageService_1.sessionStorageCache.hasOwnProperty(key)) {
            delete NSessionStorageService_1.sessionStorageCache[key];
            sessionStorage.removeItem(key);
        }
    }
    clearSessionStorage() {
        NSessionStorageService_1.sessionStorageCache = {};
        sessionStorage.clear();
    }
};
NSessionStorageService.sessionStorageCache = {};
NSessionStorageService = NSessionStorageService_1 = __decorate([
    Injectable()
], NSessionStorageService);
export { NSessionStorageService };
