// neutrinos-seed-services/esm2015/src/app/neutrinos-seed-services/n-snackbar.service.js
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
// import { MatSnackBar } from '@angular/material';
import { Subject } from 'rxjs';
import { NSystemService } from './n-system.service';
import { NPubSubService } from './n-pubSub.service';
let NSnackbarService = class NSnackbarService {
    constructor(nPubSubService) {
        this.nPubSubService = nPubSubService;
        this.snackService = new Subject();
        if (NSystemService.getInstance().getVal('useDefaultExceptionUI')) {
            this.snackBarListner = this.nPubSubService.$sub('openSnackBar', (msg) => {
                this.openSnackBar(msg);
            });
        }
    }
    get() {
        return this.snackService.asObservable();
    }
    add(data) {
        this.snackService.next({
            action: 'add',
            data: data
        });
    }
    remove(id) {
        this.snackService.next({ action: 'remove', id: id });
    }
    clear() {
        this.snackService.next({ action: 'clear' });
    }
    openSnackBar(message, timeout) {
        this.add({
            msg: message,
            timeout: timeout || 5000,
            action: {
                text: 'Close',
                onClick: (snack) => {
                },
            },
            onAdd: (snack) => {
            },
            onRemove: (snack) => {
            }
        });
    }
};
NSnackbarService.ctorParameters = () => [
    { type: NPubSubService }
];
NSnackbarService = __decorate([
    Injectable()
], NSnackbarService);
export { NSnackbarService };
