// neutrinos-module/esm2015/src/app/neutrinos-module/neutrinos-file/directives/n-barcode.directive.js
import { __decorate } from "tslib";
import { Directive, Input, Output, EventEmitter } from '@angular/core';
import { HostListener } from '@angular/core';
import { NFileIOService } from '../services/n-fileIO.service';
let NBarcodeDirective = class NBarcodeDirective {
    constructor(fsv) {
        this.fsv = fsv;
        this.onsuccess = new EventEmitter();
        this.onerror = new EventEmitter();
    }
    methodToHandleMouseClickAction() {
        this.getBarcode()
            .then(sucess => this.onsuccess.emit(sucess))
            .catch(error => this.onerror.emit(error));
    }
    getBarcode() {
        return new Promise((resolve, reject) => {
            this.fsv.getBarcode(this.barcodeOptions).then(res => {
                return resolve(res);
            }).catch(err => reject(err));
        });
    }
};
NBarcodeDirective.ctorParameters = () => [
    { type: NFileIOService }
];
__decorate([
    Input()
], NBarcodeDirective.prototype, "barcodeOptions", void 0);
__decorate([
    Output()
], NBarcodeDirective.prototype, "onsuccess", void 0);
__decorate([
    Output()
], NBarcodeDirective.prototype, "onerror", void 0);
__decorate([
    HostListener('click')
], NBarcodeDirective.prototype, "methodToHandleMouseClickAction", null);
NBarcodeDirective = __decorate([
    Directive({
        selector: '[n-barcode]'
    })
], NBarcodeDirective);
export { NBarcodeDirective };
