// neutrinos-module/esm2015/src/app/neutrinos-module/neutrinos-file/directives/n-ocr.directive.js
import { __decorate } from "tslib";
import { Directive, Input, Output, EventEmitter } from '@angular/core';
import { HostListener } from '@angular/core';
import { NFileIOService } from '../services/n-fileIO.service';
let NOcrDirective = class NOcrDirective {
    constructor(fsv) {
        this.fsv = fsv;
        this.onsuccess = new EventEmitter();
        this.onerror = new EventEmitter();
    }
    methodToHandleMouseClickAction() {
        this.getOcr()
            .then(sucess => this.onsuccess.emit(sucess))
            .catch(error => this.onerror.emit(error));
    }
    getOcr() {
        return new Promise((resolve, reject) => {
            this.fsv.getOcr(this.ocrOptions).then(res => {
                return resolve(res);
            }).catch(err => reject(err));
        });
    }
};
NOcrDirective.ctorParameters = () => [
    { type: NFileIOService }
];
__decorate([
    Input()
], NOcrDirective.prototype, "ocrOptions", void 0);
__decorate([
    Output()
], NOcrDirective.prototype, "onsuccess", void 0);
__decorate([
    Output()
], NOcrDirective.prototype, "onerror", void 0);
__decorate([
    HostListener('click')
], NOcrDirective.prototype, "methodToHandleMouseClickAction", null);
NOcrDirective = __decorate([
    Directive({
        selector: '[n-ocr]'
    })
], NOcrDirective);
export { NOcrDirective };
