// neutrinos-module/esm5/src/app/neutrinos-module/neutrinos-file/directives/n-camera.directive.js
import { __decorate } from "tslib";
import { Directive, Input, Output, EventEmitter } from '@angular/core';
import { HostListener } from '@angular/core';
import { NFileIOService } from '../services/n-fileIO.service';
var NCameraDirective = /** @class */ (function () {
    function NCameraDirective(fsv) {
        this.fsv = fsv;
        this.onsuccess = new EventEmitter();
        this.onerror = new EventEmitter();
    }
    NCameraDirective.prototype.methodToHandleMouseClickAction = function () {
        var _this = this;
        this.getPicture()
            .then(function (sucess) { return _this.onsuccess.emit(sucess); })
            .catch(function (error) { return _this.onerror.emit(error); });
    };
    NCameraDirective.prototype.getPicture = function () {
        var _this = this;
        return new Promise(function (resolve, reject) {
            _this.fsv.getPicture(_this.cameraOptions).then(function (res) {
                var options = {
                    'formData': res,
                    'entityName': _this.cameraOptions.entityName,
                    'metadata': _this.cameraOptions.metadata
                };
                _this.fsv.upload(options)
                    .then(function (uri) { return resolve(uri); })
                    .catch(function (err) { return reject(err); });
            }).catch(function (err) { return reject(err); });
        });
    };
    NCameraDirective.ctorParameters = function () { return [
        { type: NFileIOService }
    ]; };
    __decorate([
        Input()
    ], NCameraDirective.prototype, "cameraOptions", void 0);
    __decorate([
        Output()
    ], NCameraDirective.prototype, "onsuccess", void 0);
    __decorate([
        Output()
    ], NCameraDirective.prototype, "onerror", void 0);
    __decorate([
        HostListener('click')
    ], NCameraDirective.prototype, "methodToHandleMouseClickAction", null);
    NCameraDirective = __decorate([
        Directive({
            selector: '[n-camera]'
        })
    ], NCameraDirective);
    return NCameraDirective;
}());
export { NCameraDirective };
