// neutrinos-module/esm5/src/app/neutrinos-module/neutrinos-file/directives/n-ocr.directive.js
import { __decorate } from "tslib";
import { Directive, Input, Output, EventEmitter } from '@angular/core';
import { HostListener } from '@angular/core';
import { NFileIOService } from '../services/n-fileIO.service';
var NOcrDirective = /** @class */ (function () {
    function NOcrDirective(fsv) {
        this.fsv = fsv;
        this.onsuccess = new EventEmitter();
        this.onerror = new EventEmitter();
    }
    NOcrDirective.prototype.methodToHandleMouseClickAction = function () {
        var _this = this;
        this.getOcr()
            .then(function (sucess) { return _this.onsuccess.emit(sucess); })
            .catch(function (error) { return _this.onerror.emit(error); });
    };
    NOcrDirective.prototype.getOcr = function () {
        var _this = this;
        return new Promise(function (resolve, reject) {
            _this.fsv.getOcr(_this.ocrOptions).then(function (res) {
                return resolve(res);
            }).catch(function (err) { return reject(err); });
        });
    };
    NOcrDirective.ctorParameters = function () { return [
        { type: NFileIOService }
    ]; };
    __decorate([
        Input()
    ], NOcrDirective.prototype, "ocrOptions", void 0);
    __decorate([
        Output()
    ], NOcrDirective.prototype, "onsuccess", void 0);
    __decorate([
        Output()
    ], NOcrDirective.prototype, "onerror", void 0);
    __decorate([
        HostListener('click')
    ], NOcrDirective.prototype, "methodToHandleMouseClickAction", null);
    NOcrDirective = __decorate([
        Directive({
            selector: '[n-ocr]'
        })
    ], NOcrDirective);
    return NOcrDirective;
}());
export { NOcrDirective };
