// neutrinos-module/esm2015/src/app/neutrinos-module/neutrinos-components/nAlertComponent/n-alert.component.js
var NAlertComponent_1;
import { __decorate } from "tslib";
import { Component } from '@angular/core';
import { MatDialogRef } from '@angular/material/dialog';
let NAlertComponent = NAlertComponent_1 = class NAlertComponent {
    constructor(dialogRef) {
        this.dialogRef = dialogRef;
        this.messageContent = NAlertComponent_1.message;
        this.titleContent = NAlertComponent_1.title;
    }
};
NAlertComponent.title = '';
NAlertComponent.message = '';
NAlertComponent.ctorParameters = () => [
    { type: MatDialogRef }
];
NAlertComponent = NAlertComponent_1 = __decorate([
    Component({
        selector: 'n-alert-dialog',
        template: "\r\n<div class=\"alertContainer\" fxLayout=\"column\">\r\n  <h4 mat-dialog-title>{{titleContent}}</h4>\r\n  <span class=\"msgContainer\" >{{messageContent}}</span>\r\n  <button mat-raised-button (click)=\"dialogRef.close()\">Close</button>\r\n</div>\r\n",
        styles: [".alertContainer{height:auto;width:auto}.msgContainer{margin-bottom:.5em}"]
    })
], NAlertComponent);
export { NAlertComponent };
