// neutrinos-module/esm5/src/app/neutrinos-module/neutrinos-file/fileUpload/n-fileUpload.component.js
import { __decorate } from "tslib";
import { Component, Input, Output, EventEmitter, ViewChild, Renderer2 } from '@angular/core';
import { NFileIOService } from '../services/n-fileIO.service';
var NFileUploadComponent = /** @class */ (function () {
    function NFileUploadComponent(fileIOService, _renderer) {
        this.fileIOService = fileIOService;
        this._renderer = _renderer;
        this.onsuccess = new EventEmitter();
        this.onerror = new EventEmitter();
    }
    /**
     * Method executed when upload button is clicked.
     */
    NFileUploadComponent.prototype.handleUpload = function () {
        var _this = this;
        this.isDone = false;
        if (this.file && this.uploadOptions && this.uploadOptions.entityName && this.uploadOptions.metadata) {
            this.fileIOService.upload({ files: this.file, entityName: this.uploadOptions.entityName, metadata: this.uploadOptions.metadata })
                .then(function (res) {
                _this.isDone = true;
                _this.onsuccess.emit(res);
            })
                .catch(function (err) { return _this.onerror.emit(err); });
        }
        else {
            this.onerror.emit(new Error('Upload options missing'));
        }
    };
    /**
     * Method executed when a file is selected.
     */
    NFileUploadComponent.prototype.handleSelect = function (fileInput) {
        if (fileInput.target.files && fileInput.target.files[0]) {
            this.file = fileInput.target.files[0];
            this.fileName = this.file.name;
        }
    };
    /**
     * Methods executed when cancel button is clicked.
     * Clears files.
     */
    NFileUploadComponent.prototype.cancel = function () {
        this.file = null;
        this.fileName = null;
        this.isDone = false;
        // check if the file input is rendered before clearing it
        if (this.fileInput) {
            this._renderer.setProperty(this.fileInput, 'value', '');
        }
    };
    NFileUploadComponent.prototype.disableFileUploadButton = function () {
        if (typeof this.disableButton === 'boolean') {
            return this.disableButton;
        }
        return this.disableButton === 'disabled' || this.disableButton === 'true' ? true : false;
    };
    NFileUploadComponent.ctorParameters = function () { return [
        { type: NFileIOService },
        { type: Renderer2 }
    ]; };
    __decorate([
        ViewChild('fileInput', { static: false })
    ], NFileUploadComponent.prototype, "fileInput", void 0);
    __decorate([
        Input()
    ], NFileUploadComponent.prototype, "uploadOptions", void 0);
    __decorate([
        Input('disabled')
    ], NFileUploadComponent.prototype, "disableButton", void 0);
    __decorate([
        Output()
    ], NFileUploadComponent.prototype, "onsuccess", void 0);
    __decorate([
        Output()
    ], NFileUploadComponent.prototype, "onerror", void 0);
    NFileUploadComponent = __decorate([
        Component({
            selector: 'n-file-upload',
            template: "<div>\r\n  <button mat-raised-button class=\"bh-file-upload\" (click)=\"fileInput.click()\" color=\"primary\"\r\n    *ngIf=\"!fileName\" [disabled]=\"disableFileUploadButton()\">\r\n    <mat-icon>attach_file</mat-icon>\r\n    <span>Choose a file...</span>\r\n    <input #fileInput type=\"file\" (change)=\"handleSelect($event)\" style=\"display:none;\" />\r\n  </button>\r\n  <div *ngIf=\"fileName\">\r\n    <button mat-raised-button color=\"primary\" (click)=\"handleUpload()\">\r\n      <mat-icon>file_upload</mat-icon>\r\n      <span>{{fileName}}</span>\r\n    </button>\r\n    <button *ngIf=\"!isDone\" mat-icon-button color=\"warn\" (click)=\"cancel()\" class=\"bh-file-upload-cancel\">\r\n      <mat-icon>cancel</mat-icon>\r\n    </button>\r\n    <button *ngIf=\"isDone\" mat-icon-button (click)=\"cancel()\" class=\"bh-file-upload-done\">\r\n      <mat-icon>done</mat-icon>\r\n    </button>\r\n  </div>\r\n</div>\r\n",
            styles: [".bh-file-upload{padding-left:8px;padding-right:8px}.bh-file-upload-cancel{height:24px;width:24px;position:relative;top:24px;left:-12px}::ng-deep [dir=rtl] .bh-file-upload-cancel{right:-12px;left:0}.bh-file-upload-cancel mat-icon{border-radius:12px;vertical-align:baseline}.bh-file-upload-done{height:24px;width:24px;position:relative;top:20px;left:-13px}::ng-deep [dir=rtl] .bh-file-upload-done{right:-12px;left:0}.bh-file-upload-done mat-icon{border-radius:12px;vertical-align:baseline}"]
        })
    ], NFileUploadComponent);
    return NFileUploadComponent;
}());
export { NFileUploadComponent };
